janus_server/src/bin/aggregation_job_creator.rs

```diff
@@ -0,0 +1,705 @@
+use janus_server::message::{AggregationJobId, Nonce, Role, TaskId, Time};
+use janus_server::task::Task;
+use janus_server::time::{Clock, RealClock};
+use janus_server::trace::install_trace_subscriber;
+use prio::codec::Encode;
+use prio::vdaf;
+use prio::vdaf::prio3::{Prio3Aes128Count, Prio3Aes128Histogram, Prio3Aes128Sum};
+use rand::{thread_rng, Rng};
+use std::collections::HashMap;
+use std::fmt::Formatter;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::Duration;
+use std::{fmt, fmt::Debug, fs};
+use structopt::StructOpt;
+use tokio::sync::oneshot::{self, Receiver, Sender};
+use tokio::time::{Instant, MissedTickBehavior};
+use tokio::{select, time};
+use tracing::{debug, error, info};
+
+#[derive(StructOpt)]
+#[structopt(
+    name = "janus-aggregation-job-creator",
+    about = "Janus aggregation job creator",
+    rename_all = "kebab-case",
+    version = env!("CARGO_PKG_VERSION"),
+)]
+struct Options {
+    /// Path to configuration YAML.
+    #[structopt(
+        long,
+        env = "CONFIG_FILE",
+        parse(from_os_str),
+        takes_value = true,
+        required(true),
+        help = "path to configuration file"
+    )]
+    config_file: PathBuf,
+
+    /// Password for the PostgreSQL database connection. If specified, must not be specified in the
+    /// connection string.
+    #[structopt(long, env = "PGPASSWORD", help = "PostgreSQL password")]
+    database_password: Option<String>,
+
+    /// Datastore encryption keys.
+    #[structopt(
+        long,
+        env = "DATASTORE_KEYS",
+        takes_value = true,
+        use_delimiter = true,
+        required(true),
+        help = "datastore encryption keys, encoded in base64 then comma-separated"
+    )]
+    datastore_keys: Vec<String>,
+}
+
+impl Debug for Options {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Options")
+            .field("config_file", &self.config_file)
+            .finish()
+    }
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    // Read arguments, read & parse config.
+    let options = Options::from_args();
+    let config: AggregationJobCreatorConfig = {
+        let config_content = fs::read_to_string(&options.config_file)
+            .with_context(|| format!("couldn't read config file {:?}", options.config_file))?;
+        serde_yaml::from_str(&config_content)
+            .with_context(|| format!("couldn't parse config file {:?}", options.config_file))?
+    };
+    install_trace_subscriber(&config.logging_config)
+        .context("couldn't install tracing subscriber")?;
+
+    info!(?options, ?config, "Starting aggregation job creator");
+
+    // Connect to database.
+    let datastore = datastore(
+        config.database,
+        options.database_password,
+        options.datastore_keys,
+    )
+    .context("couldn't connect to database")?;
+
+    // Start creating aggregation jobs.
+    Arc::new(AggregationJobCreator {
```

Comment:
The technically-correct-but-unhelpful answer is that it allows us to call `AggregationJobCreator::run()`, which receives `self` as an `Arc<Self>`.

Of course, this leads to the question of `run` takes an `Arc<Self>`. This is to allow `run` to spawn tasks (which run `AggregationJobCreator::run_for_task`), handing each task a reference to the `AggregationJobCreator` by cloning the `Arc` instead of the entire `AggregationJobCreator` -- we couldn't do this if `self` was a `&Self` due to lifetime issues, since the tasks run independently from `run`. This allows `run_for_task` and methods it calls to take a `&Self` rather than an `Arc<Self>`.

The other strategy, I suppose, would be to wrap "large" or otherwise unclonable pieces of the structure (e.g. the `Datastore`) in an `Arc`, rather than the entire structure. This is the strategy that `Aggregator` takes. But this would potentially lead to more reference-count juggling, and more importantly would mean we'd need to clone/copy each part of the `AggregationJobCreator` struct we needed to pass down to methods we call rather than passing down a reference to the entire `AggregationJobCreator` -- this would be painful since there are a number of configuration parameters, it's very nice to be able to pass them all along in a single parameter.

Why doesn't `Aggregator` use a similar strategy? Maybe it should, though it has fewer "configuration parameters" so for now managing an `Arc<Datastore>` is just as good.